headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/spel/SpelDefinitionProvider.java

```diff
@@ -0,0 +1,337 @@
+import org.springframework.expression.spel.ast.PropertyOrFieldReference;
+import org.springframework.expression.spel.ast.TypeReference;
+import org.springframework.expression.spel.standard.SpelExpression;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.ide.vscode.boot.index.SpringMetamodelIndex;
+import org.springframework.ide.vscode.boot.java.Annotations;
+import org.springframework.ide.vscode.boot.java.IJavaDefinitionProvider;
+import org.springframework.ide.vscode.boot.java.links.SourceLinks;
+import org.springframework.ide.vscode.boot.java.spel.AnnotationParamSpelExtractor.Snippet;
+import org.springframework.ide.vscode.boot.java.utils.ASTUtils;
+import org.springframework.ide.vscode.boot.java.utils.CompilationUnitCache;
+import org.springframework.ide.vscode.commons.java.IJavaProject;
+import org.springframework.ide.vscode.commons.protocol.spring.Bean;
+import org.springframework.ide.vscode.commons.util.BadLocationException;
+import org.springframework.ide.vscode.commons.util.text.DocumentRegion;
+import org.springframework.ide.vscode.commons.util.text.LanguageId;
+import org.springframework.ide.vscode.commons.util.text.TextDocument;
+import org.springframework.ide.vscode.parser.spel.SpelLexer;
+import org.springframework.ide.vscode.parser.spel.SpelParser;
+import org.springframework.ide.vscode.parser.spel.SpelParser.BeanReferenceContext;
+import org.springframework.ide.vscode.parser.spel.SpelParserBaseListener;
+
+import reactor.util.function.Tuple2;
+import reactor.util.function.Tuples;
+
+/**
+ * @author Udayani V
+ */
+public class SpelDefinitionProvider implements IJavaDefinitionProvider {
+
+	protected static Logger logger = LoggerFactory.getLogger(SpelDefinitionProvider.class);
+
+	private final SpringMetamodelIndex springIndex;
+
+	private final CompilationUnitCache cuCache;
+
+	private final AnnotationParamSpelExtractor[] spelExtractors = AnnotationParamSpelExtractor.SPEL_EXTRACTORS;
+
+	public record TokenData(String text, int start, int end) {};
+
+	public SpelDefinitionProvider(SpringMetamodelIndex springIndex, CompilationUnitCache cuCache) {
+		this.springIndex = springIndex;
+		this.cuCache = cuCache;
+	}
+
+	@Override
+	public List<LocationLink> getDefinitions(CancelChecker cancelToken, IJavaProject project,
+			TextDocumentIdentifier docId, CompilationUnit cu, ASTNode n, int offset) {
+		if (n instanceof StringLiteral) {
+			StringLiteral valueNode = (StringLiteral) n;
+			ASTNode parent = ASTUtils.getNearestAnnotationParent(valueNode);
+
+			if (parent != null && parent instanceof Annotation) {
+				Annotation a = (Annotation) parent;
+				IAnnotationBinding binding = a.resolveAnnotationBinding();
+				if (binding != null && binding.getAnnotationType() != null
+						&& Annotations.VALUE.equals(binding.getAnnotationType().getQualifiedName())) {
+					return parseSpelAndFetchLocation(cancelToken, project, cu, offset);
```

Comment:
I think I was looking at this PR in the past and just didn't look closely enough :-\
I think I like everything except this piece. We're only interested in this literal the cursor is at and nothing else. At this point whatever `parseSpelAndFetchLocation(...)` is doing is wrong. No need to visit the CU and find all possible SPEL expressions and collect them. Instead of CU you should really visit just one single annotation.

Anyway, I'd try to extract SPEL Region from the found annotation `a` (line 101). Once you have the Region check that `offset` paremter is within that region. If yes, then:
```
			List<TokenData> beanReferenceTokens = computeTokens(snippet, offset);
							if (beanReferenceTokens != null && beanReferenceTokens.size() > 0) {
								locationLink.addAll(findLocationLinksForBeanRef(project, offset, beanReferenceTokens));
							}

							Optional<Tuple2<String, String>> result = parseAndExtractMethodClassPairFromSpel(snippet,
									offset);
							result.ifPresent(tuple -> {
								locationLink.addAll(findLocationLinksForMethodRef(tuple.getT1(), tuple.getT2(), project));
							});
```

Reply:
Yes, this makes sense. Thanks for pointing this out. Removed the unnecessary parsing logic.